libnavigation-base/src/main/java/com/mapbox/navigation/base/route/model/RouteLegProgressNavigation.kt

```diff
@@ -0,0 +1,262 @@
+     *
+     * @return a [LegStepNavigation] representing the previous step the user was on, if still on first
+     * step in route, returns null
+     */
+    fun previousStep(): LegStepNavigation? = previousStep
+
+    /**
+     * Returns the current step the user is traversing along.
+     *
+     * @return a [LegStepNavigation] representing the step the user is currently on
+     */
+    fun currentStep(): LegStepNavigation? = currentStep
+
+    /**
+     * Get the next/upcoming step immediately after the current step. If the user is on the last step
+     * on the last leg, this will return null since a next step doesn't exist.
+     *
+     * @return a [LegStepNavigation] representing the next step the user will be on.
+     */
+    fun upComingStep(): LegStepNavigation? = upComingStep
+
+    /**
+     * This will return the [LegStep] two steps ahead of the current step the user's on. If the
+     * user's current step is within 2 steps of their final destination this will return null.
+     *
+     * @return the [LegStepNavigation] after the [.upComingStep]
+     */
+    fun followOnStep(): LegStepNavigation? = followOnStep
+
+    /**
+     * Gives a [RouteStepProgressNavigation] object with information about the particular step the user
+     * is currently on.
+     *
+     * @return a [RouteStepProgressNavigation] object
+     */
+    fun currentStepProgress(): RouteStepProgressNavigation? = currentStepProgress
+
+    /**
+     * Provides a list of points that represent the current step
+     * step geometry.
+     *
+     * @return list of points representing the current step
+     */
+    fun currentStepPoints(): List<Point>? = currentStepPoints
+
+    /**
+     * Provides a list of points that represent the upcoming step
+     * step geometry.
+     *
+     * @return list of points representing the upcoming step
+     */
+    fun upcomingStepPoints(): List<Point>? = upcomingStepPoints
+
+    /**
+     * Not public since developer can access same information from [RouteProgressNavigation].
+     */
+    internal fun routeLeg(): RouteLegsNavigation? = routeLeg
+
+    internal fun stepDistanceRemaining(): Double = stepDistanceRemaining
+
+    fun toBuilder() = builder
+
+    class Builder {
+        private var stepIndex: Int = 0
+        private var fractionTraveled: Float = 0f
+        private var distanceTraveled: Double = 0.0
+        private var distanceRemaining: Double = 0.0
+        private var durationRemaining: Double = 0.0
+        private var stepDistanceRemaining: Double = 0.0
+        private var currentStepPoints: List<Point>? = null
+        private var upcomingStepPoints: List<Point>? = null
+        private var previousStep: LegStepNavigation? = null
+        private var upComingStep: LegStepNavigation? = null
+        private var followOnStep: LegStepNavigation? = null
+        private lateinit var routeLegBuilder: RouteLegsNavigation
+        private lateinit var currentStepBuilder: LegStepNavigation
```

Comment:
Had `routeLeg` earlier, but if you look at Line 160, when I check for `isInitialized`, it complains because `routeLeg` name is same for the variable and method. Any other suggestions for variable name?

Reply:
Something like `_routeLeg` could work.

Reply:
👍